README.md
# TBCMailServer

A private server for the mailbox in The Battle Cats.

It allows you to get any cat, talent orb, or item amount you want without
the ban risk of other methods.

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/fieryhenry)

## Credits

- [jamesiotio's CITM](https://github.com/jamestiotio/CITM) for the original
    hacking method and the format of the presents. It no longer works due to
    PONOS adding a signature to the server responses as well as other changes.

## Prerequisites

I won't go into detail on how to do all of the individual steps but you can
find tutorials online.

1. You will need to modify the game to always verify the
nyanko-signature of the server responses as we don't know PONOS's private key.
This can be done using [Frida](https://frida.re/) and this script:

```js
let func_name = "_ZN5Botan11PK_Verifier14verify_messageEPKhjS2_j"
// Botan::PK_Verifier::verify_message(...)

Interceptor.attach(Module.findExportByName("libnative-lib.so", func_name), {
    onLeave: function (retval) {
        retval.replace(0x1)
    }
})
```

You also need to replace the <https://nyanko-items.ponosgames.com> URL in the
libnative-lib.so files with your own server URL. This can be done by extracting
the APK using something like
[apktool](https://ibotpeaches.github.io/Apktool/). Then you can modify the
`libnative-lib.so` files in /lib using a hex editor (or notepad maybe). Then you
can repack the APK using apktool and sign it using
[apksigner](https://developer.android.com/studio/command-line/apksigner).

The URL needs to be the same length as the original URL and it needs to have
/items/ at the end with underscores padding the rest of the URL.
The underscores are used to make the URL the same length as the original URL.
The URL also needs to start with https and so you can use a service like
[Serveo](https://serveo.net/) to do the https part for you. Just follow the
instructions on the website. I recommend you to get a custom subdomain so it
doesn't change (you can use any subdomain as long as the total url length is
shorter than the original URL).

Example:
Run a command like: `ssh -R myserver:80:localhost:80 serveo.net`
You may need to setup ssh keys for the above to work.
This makes your url: `https://myserver.serveo.net` and then you would replace the
ponos url with `https://myserver.serveo.net/items/_`.

The new modding tool will make this process much easier when it is released.

Instead of using a private server, you might be able to use something like
[mitmproxy](https://mitmproxy.org/) or [Fiddler](https://www.telerik.com/fiddler)
to modify the server responses. This did not work for me as the game always
crashed but it might work for you.

## Installation

1. Install [Python](https://www.python.org/downloads/) >= 3.9
1. Run `pip install -U tbcms`

## Usage

1. Run `python -m tbcms` or `py -m tbcms` depending on
    your system.
1. Run `python -m tbcms --help` for more information.
1. Read [jamesiotio's CITM](https://github.com/jamestiotio/CITM) on how to
    format the list of presents.
1. Create a file called `presents.json` somewhere and put your presents in it.
1. Run `python -m tbcms --presents path/to/presents.json` to start
    server.

Example presents.json

```json
[
    {
        "id": 1,
        "title": "Items",
        "body": "Test Body",
        "createdAt": 1688648392,
        "items": [
            {
                "itemId": 22,
                "itemCategory": 0,
                "amount": 50,
                "title": "Catfood"
            },
            {
                "itemId": 29,
                "itemCategory": 0,
                "amount": 5,
                "title": "Platinum Ticket"
            }
        ]
    }
]
```

Note that `created` is now `createdAt`. You also don't need to fill in `country`
or `clientVersion` or `accountId` as the server will do that for you.

## Alternative Method

You can also change the public key the game uses to verify the server responses
to your own public key from a key pair you generated.

You can modify the `assets/nyanko-service-prd.pem` file in the APK to use your own
public key. You can generate a key pair using openssl:

```sh
openssl genrsa -out private.pem 4096
openssl rsa -in private.pem -pubout -out public.pem
```

The problem with this method is that you now need to intercept or use a private
server for every single request that uses the public key as now the game will
reject the official server responses from PONOS.

However, I did still manage to do all of that, but the game refused to upload
the save data to the game servers. It does a request to an aws server but it
aborts immediately after. I couldn't have changed the url of the server it
uploads to to my own server because the url is from a response of another
request with a signature that I don't know how to generate.

Another problem is that if you have a request with more than one slash in the
same place, serveo.net will respond with a 301 Moved Permanently and redirect to
the url with only one slash. But if the original request was a POST request,
the method will change to GET and break the request. If it instead used a 308
Permanent Redirect then it would keep the method as POST and it would work.
When downloading save data, PONOS accidentally has 2 slashes in the same place.
I don't know how to fix this.

I know that this is not a Flask issue because Flask responds correctly with a
308 Permanent Redirect. I think it is a problem with serveo.net.

It's probably possible to do this method but I can't be bothered to do it.

Even though this alternative method is not effective, I spent a lot of time on it so I
wanted to include it here.


CHANGELOG.md
# Changelog

## 1.0.0 - 2023-07-06

### Added

- Initial release


src/bcmbps/server.py
import json
import os
import re
import shutil
import time
from typing import Any, Optional

import flask
import jwt
import requests


def get_files_folder() -> str:
    """Get the path to the files folder

    Returns:
        str: The path to the files folder
    """
    file = os.path.realpath(__file__)
    path = os.path.dirname(file)
    path = os.path.join(path, "files")
    return path


def get_file_path(path: str) -> str:
    """Get the path to a file in the files folder

    Args:
        path (str): The path to the file

    Returns:
        str: The path to the file
    """
    path = os.path.join(get_files_folder(), path)
    return path


def get_presents() -> list[dict[str, Any]]:
    """Get the presents from the presents.json file

    Returns:
        list[dict[str, Any]]: The presents
    """
    path = get_file_path("presents.json")
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)

    item = {
        "id": 1,
        "title": "Mailbox Private Server",
        "body": "Made by fieryhenry and made possible\nby jamestiotio's original work:\nhttps://github.com/jamestiotio/CITM",
        "createdAt": int(time.time()),
        "items": [{"itemId": 22, "itemCategory": 0, "amount": 0, "title": "Catfood"}],
    }
    data.insert(
        0,
        item,
    )
    return data


def save_presents(path: str):
    """Save the presents to the presents.json file

    Args:
        path (str): The path to the presents
    """
    if path == get_file_path("presents.json"):
        return
    shutil.copy(
        path,
        get_file_path("presents.json"),
    )


app = flask.Flask(__name__)


@app.route("/items/<path:path>")
@app.route("/items/")
def items(path: str = "") -> Any:
    """Handle the items endpoint

    Args:
        path (str, optional): The path to the endpoint. Defaults to "".

    Returns:
        Any: The response
    """
    path = fix_path(path)
    if "presents/count" in path:
        return items_presents_count()
    if "presents/reception" in path:
        return items_presents_reception()
    if "presents" in path:
        return items_presents()
    return forward_request_rq(
        flask.request, "https://nyanko-items.ponosgames.com/" + path
    )


def read_auth_header() -> dict[str, Any]:
    """Read the Authorization header

    Returns:
        dict[str, Any]: The data in the header
    """
    auth_header = flask.request.headers.get("Authorization")
    if auth_header is None:
        return {}

    match = re.match(r"Bearer (?P<token>.+)", auth_header)
    if match is None:
        return {}

    token = match.group("token")
    if token is None:
        return {}

    data = jwt.decode(token, algorithms=["HS256"], options={"verify_signature": False})
    return data


def get_account_code(token: dict[str, Any]) -> str:
    """Get the account code from the token

    Args:
        token (dict[str, Any]): The token

    Returns:
        str: The account code
    """
    return token.get("accountCode", "")


def get_country_code(token: dict[str, Any]) -> str:
    """Get the country code from the token

    Args:
        token (dict[str, Any]): The token

    Returns:
        str: The country code
    """
    client_info: dict[str, Any] = token.get("clientInfo", {})
    client: dict[str, Any] = client_info.get("client", {})
    country_code: str = client.get("countryCode", "")
    return country_code


def get_version(token: dict[str, Any]) -> str:
    """Get the version from the token

    Args:
        token (dict[str, Any]): The token

    Returns:
        str: The version
    """
    client_info: dict[str, Any] = token.get("clientInfo", {})
    client: dict[str, Any] = client_info.get("client", {})
    version: str = client.get("version", "")
    return version


def items_presents_count() -> Any:
    """Handle the items/presents/count endpoint

    Returns:
        Any: The response
    """
    nonce = flask.request.args.get("nonce")
    if nonce is None:
        return "nonce is None"

    json_data = {
        "statusCode": 1,
        "nonce": nonce,
        "payload": {
            "count": len(get_presents()),
        },
        "timestamp": int(time.time()),
    }

    body = json.dumps(json_data).encode()

    headers = {
        "Content-Type": "application/json",
        "Nyanko-Signature": "A",
    }

    return flask.Response(body, status=200, headers=headers)


def items_presents() -> Any:
    """Handle the items/presents endpoint

    Returns:
        Any: The response
    """
    nonce = flask.request.args.get("nonce")
    if nonce is None:
        return "nonce is None"

    auth_header = read_auth_header()
    account_code = get_account_code(auth_header)
    country_code = get_country_code(auth_header)
    version = get_version(auth_header)

    presents = get_presents()
    for present in presents:
        present["country"] = country_code
        present["clientVersion"] = version
        present["accountId"] = account_code

    json_data = {
        "statusCode": 1,
        "nonce": nonce,
        "payload": {
            "presents": presents,
        },
        "timestamp": int(time.time()),
    }

    body = json.dumps(json_data).encode()

    headers = {
        "Content-Type": "application/json",
        "Nyanko-Signature": "A",
    }

    return flask.Response(body, status=200, headers=headers)


def items_presents_reception() -> Any:
    """Handle the items/presents/reception endpoint

    Returns:
        Any: The response
    """
    nonce = flask.request.args.get("nonce")
    if nonce is None:
        return "nonce is None"

    json_data = {
        "statusCode": 1,
        "nonce": nonce,
        "timestamp": int(time.time()),
    }

    body = json.dumps(json_data).encode()

    headers = {
        "Content-Type": "application/json",
        "Nyanko-Signature": "A",
    }

    return flask.Response(body, status=200, headers=headers)


def forward_request(
    url: str,
    args: dict[str, Any],
    method: str,
    content: bytes,
    headers: dict[str, Any],
) -> flask.Response:
    """Forward a request to another server

    Args:
        url (str): URL to forward to
        args (dict[str, Any]): HTTP arguments
        method (str): HTTP method
        content (bytes): HTTP content
        headers (dict[str, Any]): HTTP headers

    Returns:
        flask.Response: The response
    """
    real_response = requests.request(
        method,
        url,
        params=args,
        headers=headers,
        data=content,
        timeout=5,
    )
    real_headers = dict(real_response.headers)
    real_content = real_response.content
    real_response_code = real_response.status_code
    real_headers.pop("Transfer-Encoding", None)

    real_headers["Nyanko-Signature"] = "A"
    return flask.Response(real_content, status=real_response_code, headers=real_headers)


def forward_request_rq(request: flask.Request, url: str) -> flask.Response:
    """Forward a request to another server

    Args:
        request (flask.Request): The request to forward
        url (str): URL to forward to

    Returns:
        flask.Response: The response
    """
    args = request.args
    method = request.method
    content = request.get_data()
    headers = request.headers

    return forward_request(url, args, method, content, headers)


def fix_path(path: str) -> str:
    """Fix the path, removing padding underscores and glitched stack string stuff?

    Args:
        path (str): The path to fix

    Returns:
        str: The fixed path
    """
    if not path.startswith("/"):
        path = "/" + path
    if not path.endswith("/"):
        path = path + "/"
    path = re.sub(r"/_+", "/", path)
    path = path.replace("/.com/", "/")  # idk why this happens
    path = path.replace("/om/", "/")  # idk why this happens

    return path


def run(
    debug: bool = False,
    port: int = 80,
    host: str = "0.0.0.0",
    presents_path: Optional[str] = None,
) -> None:
    """Run the server"""
    if presents_path is not None:
        save_presents(presents_path)
    app.run(debug=debug, host=host, port=port)
